microprofile/WFLY-12609_MicroProfile_JWT_1_1.adoc

```diff
@@ -0,0 +1,109 @@
+= WFLY-12609 MicroProfile JWT 1.1 Support
+:author:            Darran Lofthouse
+:email:             darran.lofthouse@redhat.com
+:toc:               left
+:icons:             font
+:idprefix:
+:idseparator:       -
+
+== Overview
+
+This proposal is in relation to adding support to WildFly for the MicroProfile JWT 1.1 specification.
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.jboss.org/browse/WFLY-12609[WFLY-12609]
+
+=== Related Issues
+
+* https://issues.jboss.org/browse/EAP7-1346[EAP7-1346]
+
+=== Dev Contacts
+
+* mailto:{email}[{author}]
+
+=== QE Contacts
+
+=== Testing By
+// Put an x in the relevant field to indicate if testing will be done by Engineering or QE. 
+// Discuss with QE during the Kickoff state to decide this
+[ ] Engineering
+
+[ ] QE
+
+=== Affected Projects or Components
+
+The primary integration will be happening by the addition of a new subsystem `microprofile=jwt` within https://github.com/wildfly/wildfly[WildFly].
+
+Additional integration may be added within https://github.com/wildfly-security/wildfly-elytron[WildFly Elytron] to bridge into the application server's security framework.
+
+=== Other Interested Projects
+
+The integration will also be making use of https://github.com/smallrye/smallrye-jwt[SmallRye JWT] as well as pulling in the https://github.com/eclipse/microprofile-jwt-auth[Eclipse MicroProfile JWT] API library.
+
+== Requirements
+
+=== Hard Requirements
+
+Primarily this enhancement will be delivered using a new subsystem `microprofile-jwt` which will be added to WildFly.  The subsystem will install a DeploymentUnitProcessor which will detect if MicroProfile JWT is required and activate a suitable security configuration as described within this proposal.
+
+At the outset activation will only occur in the following circumstances: -
+
+ * A `@org.eclipse.microprofile.auth.LoginConfig` annotation is detected with an `auth-method` of `MP-JWT`.
+ * An auth method of `MP-JWT` is specified in any web.xml.
+
+This activation scope is deliberately constrained to allow us to incrementally increase the scope at a later point. 
+
+At the point activation occurs the deployment will not make use of a configured `SecurityDomain` instead a "Virtual" `SecurityDomain` will be created to meet the requirements of the deployment and will be used across all sub-deployments.  The common `SecurityDomain` will allow for identity propagation across all sub-deployments whilst the virtual nature of the domain will make it unavailable for propagation across unrelated deployments allowing for deployment level isolation.
+
+Activation will add module dependencies as required on SmallRye JWT, Elytron JWT, and the MicroProfile JWT API.
+
+Activation will also ensure activation of the SmallRye JWT CDI extension as well as any Elytron JWT CDI extension if deemed required and any Elytron specific overrides.
```

Comment:
This subsystem is operating on Elytron APIs ONLY there will be no PicketBox variant.

Reply:
That's a bit confusing atm. PicketBox is still available and can be the default. In such case JWT functionability wont be available ? 

I thought this will be transparent for JWT usage / backend impl agnostic. 
Only if users will want to use something "advanced" they will need to use Elytron.

Reply:
From the perspective of deployments there is actually no notion of a default - for all deployments appropriate security resources need to be defined before a deployment is deployed to the application server and the administrator can choose to use either Elytron or the legacy PicketBox resources.

This specific enhancement is adding support for the security configuration to be provided exclusively using the MicroProfile Config.

For this reason there is no association with managed security resources instead everything is created dynamically and internally to meet the needs of the deployment.

From the perspective of the developer the security framework in use is entirely internal to the application server - it would only be if they wanted to make use of vendor specific APIs that they would need to use Elytron APIs.

If it wasn't for the server exposing it's management interfaces the Elytron subsystem and Security subsystem almost become optional - it would not take too many steps to be able to remove these completely leaving a server distribution to just support JWT.